lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -4724,6 +4724,106 @@ FIRRTLType RefSubOp::inferReturnType(ValueRange operands,
       loc, "ref.sub op requires a RefType of vector or bundle base type");
 }
 
+//===----------------------------------------------------------------------===//
+// Optional Group Operations
+//===----------------------------------------------------------------------===//
+
+LogicalResult GroupOp::verify() {
+  auto groupName = getGroupName();
+  auto *parentOp = (*this)->getParentOp();
+
+  // Verify the correctness of the symbol reference.  Only verify that this
+  // group makes sense in its parent module or group.
+  auto nestedReferences = groupName.getNestedReferences();
+  if (nestedReferences.empty()) {
+    if (!isa<FModuleOp>(parentOp)) {
+      auto diag = emitOpError() << "has an un-nested group symbol, but does "
+                                   "not have a 'firrtl.module' op as a parent";
+      return diag.attachNote(parentOp->getLoc())
+             << "illegal parent op defined here";
+    }
+  } else {
+    auto parentGroup = dyn_cast<GroupOp>(parentOp);
+    if (!parentGroup) {
+      auto diag = emitOpError()
+                  << "has a nested group symbol, but does not have a '"
+                  << getOperationName() << "' op as a parent'";
+      return diag.attachNote(parentOp->getLoc())
+             << "illegal parent op defined here";
+    }
+    auto parentGroupName = parentGroup.getGroupName();
+    if (parentGroupName.getRootReference() != groupName.getRootReference() ||
+        parentGroupName.getNestedReferences() !=
+            groupName.getNestedReferences().drop_back()) {
+      auto diag = emitOpError() << "is nested under an illegal group";
+      return diag.attachNote(parentGroup->getLoc())
+             << "illegal parent group defined here";
+    }
+  }
+
+  // Verify the body of the region.
+  Block *body = getBody(0);
+  bool failed = false;
+  getBody(0)->walk([&](Operation *op) {
+    // Skip nested groups.  Those will be verified separately.
+    if (isa<GroupOp>(op))
+      return WalkResult::skip();
```

Comment:
I think the walk order is post order by default, so it's necessary to manually specify the order to get the intended behavior
```suggestion
  getBody(0)->walk<WalkOrder:PreOrder>([&](Operation *op) {
    // Skip nested groups.  Those will be verified separately.
    if (isa<GroupOp>(op))
      return WalkResult::skip();
```

Reply:
I don't think there's any functional problem here. However, this is a good thought such that it would be better to report errors _in IR order_ as opposed to from nested whens outwards. 👍 I'll change this.

Reply:
The error reporting isn't a problem. I was confused. Yes, this seems better to go `PreOrder` here. 👍